File: typescript/server/test/identity-intake.test.ts
/**
 * The `identity_intake` Rich Interaction kind — validator + kind wiring + the
 * shared conformance fixtures.
 *
 * The TS parity of the Rust `identity_intake.rs` unit tests: email/phone
 * normalization, the one-pass required/optional + per-field validation, the
 * raise-tool argument contract (`parseFields`), the kind's
 * `parseRequest`/`validate`/`fallbackDirective`, and the shared `identity_intake`
 * fixtures in `spec/conformance/fixtures.json` (so a drift in any one server is
 * caught here).
 */
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { dirname, join } from 'node:path';
import { describe, expect, it } from 'vitest';

import { IdentityIntakeKind, normalizeEmail, normalizePhoneE164, parseFields, validateIntake, type IntakeField, type IntakeValues } from '../src/identityIntake.js';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SPEC_DIR = join(__dirname, '..', '..', '..', 'spec');
const fixtures = JSON.parse(readFileSync(join(SPEC_DIR, 'conformance', 'fixtures.json'), 'utf8')) as Record<string, { instance: Record<string, unknown> } | string>;
function fixture(name: string): Record<string, unknown> {
    const entry = fixtures[name];
    if (!entry || typeof entry === 'string') throw new Error(`missing fixture ${name}`);
    return entry.instance;
}

function field(key: IntakeField['key'], required: boolean): IntakeField {
    return { key, required };
}

describe('normalizeEmail', () => {
    it('lowercases the domain, preserves the local case', () => {
        expect(normalizeEmail('Alice@Example.COM')).toBe('Alice@example.com');
    });
    it('rejects malformed addresses', () => {
        for (const bad of ['', 'no-at', '@x.com', 'a@b', 'a@.com', 'a@b.', 'a b@c.com', 'a@b@c.com']) {
            expect(normalizeEmail(bad), `${bad} should be rejected`).toBeUndefined();
        }
    });
});

describe('normalizePhoneE164', () => {
    it('normalizes NANP and keeps country-coded numbers', () => {
        expect(normalizePhoneE164('+1 (555) 123-4567')).toBe('+15551234567');
        expect(normalizePhoneE164('555.123.4567')).toBe('+15551234567'); // bare 10-digit
        expect(normalizePhoneE164('1 555 123 4567')).toBe('+15551234567'); // 1-prefixed 11-digit
        expect(normalizePhoneE164('+447911123456')).toBe('+447911123456'); // non-NANP with country code
    });
    it('rejects unparseable numbers', () => {
        for (const bad of ['', 'abc', '+0123456789', '12345', '+1234567890123456']) {
            expect(normalizePhoneE164(bad), `${bad} should be rejected`).toBeUndefined();
        }
    });
});

describe('validateIntake', () => {
    it('a required field missing (or blank) is an error', () => {
        const fields = [field('email', true), field('name', false)];
        const err = validateIntake(fields, {});
        expect(err.ok).toBe(false);
        if (!err.ok) {
            expect(err.errors).toHaveLength(1);
            expect(err.errors[0]!.field).toBe('email');
        }
        // Blank counts as missing.
        expect(validateIntake(fields, { email: '   ' }).ok).toBe(false);
    });

    it('an optional field left absent is fine', () => {
        const out = validateIntake([field('name', false), field('email', false)], { name: 'Bob' });
        expect(out.ok).toBe(true);
        if (out.ok) expect(out.values).toEqual({ name: 'Bob' });
    });

    it('a valid submit trims + normalizes every field', () => {
        const fields = [field('email', true), field('phone', false)];
        const out = validateIntake(fields, { name: '  Alice Example  ', email: 'alice@Example.com', phone: '(555) 123-4567' });
        expect(out.ok).toBe(true);
        if (out.ok) {
            expect(out.values.name).toBe('Alice Example');
            expect(out.values.email).toBe('alice@example.com');
            expect(out.values.phone).toBe('+15551234567');
        }
    });

    it('a bad email → a per-field email error', () => {
        const out = validateIntake([field('email', true)], { email: 'not-an-email' });
        expect(out.ok).toBe(false);
        if (!out.ok) {
            expect(out.errors).toHaveLength(1);
            expect(out.errors[0]!.field).toBe('email');
            expect(out.errors[0]!.message).toContain('valid email');
        }
    });

    it('reports missing name + bad email + bad phone in one pass', () => {
        const out = validateIntake([field('name', true)], { email: 'not-an-email', phone: 'nope' });
        expect(out.ok).toBe(false);
        if (!out.ok) expect(out.errors).toHaveLength(3);
    });

    it('keeps a volunteered field that was not requested', () => {
        const out = validateIntake([field('email', true)], { email: 'a@b.co', phone: '+15551234567' });
        expect(out.ok).toBe(true);
        if (out.ok) expect(out.values.phone).toBe('+15551234567');
    });
});

describe('parseFields — the raise-tool contract', () => {
    it('accepts the shorthand string form (required: true) and the object form', () => {
        const shorthand = parseFields(['email', 'name']);
        expect(shorthand).toEqual([{ key: 'email', required: true }, { key: 'name', required: true }]);
        const structured = parseFields([{ key: 'phone', required: false, label: 'Mobile' }]);
        expect(structured).toEqual([{ key: 'phone', required: false, label: 'Mobile' }]);
    });
    it('rejects an empty array, a non-array, and an unknown key', () => {
        expect(() => parseFields([])).toThrow(/at least one/);
        expect(() => parseFields('nope')).toThrow(/must be an array/);
        expect(() => parseFields(['address'])).toThrow(/unknown intake field/);
    });
});

describe('IdentityIntakeKind', () => {
    it('exposes the reference identity + raise-tool surface', () => {
        const kind = new IdentityIntakeKind();
        expect(kind.kind).toBe('identity_intake');
        expect(kind.capability).toBe('identity_form');
        expect(kind.toolSchema().name).toBe('request_identity_intake');
    });

    it('parseRequest canonicalizes fields + reason', () => {
        const req = new IdentityIntakeKind().parseRequest({ fields: ['email', { key: 'phone', required: false }], reason: 'to send you the quote' });
        expect(req.kind).toBe('identity_intake');
        expect(req.reason).toBe('to send you the quote');
        expect((req.spec as { fields: IntakeField[] }).fields).toEqual([{ key: 'email', required: true }, { key: 'phone', required: false }]);
    });

    it('fallbackDirective lists the fields and points at submit_interaction', () => {
        const kind = new IdentityIntakeKind();
        const req = kind.parseRequest({ fields: ['name', 'email'], reason: 'to follow up' });
        const directive = kind.fallbackDirective(req.spec, 'to follow up');
        expect(directive).toContain('name, email');
        expect(directive).toContain('submit_interaction');
    });

    it('validate rejects an all-empty submit', () => {
        const out = new IdentityIntakeKind().validate({ fields: [{ key: 'email', required: true }] }, {});
        expect(out.ok).toBe(false);
        if (!out.ok) expect(out.errors[0]!.field).toBe('values');
    });
});

describe('the shared identity_intake fixtures', () => {
    it('validate(identity_intake_spec, identity_intake_values) produces identity_intake_payload.values', () => {
        const spec = fixture('identity_intake_spec');
        const values = fixture('identity_intake_values') as IntakeValues;
        const payload = fixture('identity_intake_payload');
        const out = new IdentityIntakeKind().validate(spec, values);
        expect(out.ok).toBe(true);
        if (out.ok) expect(out.values).toEqual((payload as { values: unknown }).values);
    });
});

File: go/server/identity_intake_test.go
package server

import (
	"encoding/json"
	"testing"
)

// Unit tests for the identity_intake Rich Interaction kind — the Go port of the Rust
// rust/smooth-operator/src/identity_intake.rs tests, plus validation against the shared
// conformance fixtures (spec/conformance/fixtures.json), so the Go validator agrees with
// the Rust reference on the same wire shapes.

func ifield(key string, required bool) IntakeField {
	return IntakeField{Key: key, Required: required}
}

func TestNormalizeEmailShapes(t *testing.T) {
	if got, ok := normalizeEmail("Alice@Example.COM"); !ok || got != "Alice@example.com" {
		t.Fatalf("domain lowercased/local preserved: got %q ok=%v", got, ok)
	}
	for _, bad := range []string{"", "no-at", "@x.com", "a@b", "a@.com", "a@b.", "a b@c.com", "a@b@c.com"} {
		if _, ok := normalizeEmail(bad); ok {
			t.Errorf("%q should be rejected", bad)
		}
	}
}

func TestNormalizePhoneE164(t *testing.T) {
	cases := map[string]string{
		"+1 (555) 123-4567": "+15551234567",
		"555.123.4567":      "+15551234567",  // bare 10-digit NANP
		"1 555 123 4567":    "+15551234567",  // 1-prefixed 11-digit NANP
		"+447911123456":     "+447911123456", // non-NANP with country code
	}
	for in, want := range cases {
		if got, ok := normalizePhoneE164(in); !ok || got != want {
			t.Errorf("normalizePhoneE164(%q) = %q,%v want %q", in, got, ok, want)
		}
	}
	for _, bad := range []string{"", "abc", "+0123456789", "12345", "+1234567890123456"} {
		if _, ok := normalizePhoneE164(bad); ok {
			t.Errorf("%q should be rejected", bad)
		}
	}
}

func TestValidateIntakeRequiredMissingIsError(t *testing.T) {
	fields := []IntakeField{ifield("email", true), ifield("name", false)}
	_, errs := validateIntake(fields, IntakeValues{})
	if len(errs) != 1 || errs[0].Field != "email" {
		t.Fatalf("expected one email-required error, got %+v", errs)
	}
	// Blank counts as missing.
	if _, errs := validateIntake(fields, IntakeValues{Email: "   "}); len(errs) == 0 {
		t.Fatal("blank required email should error")
	}
}

func TestValidateIntakeValidNormalizes(t *testing.T) {
	fields := []IntakeField{ifield("email", true), ifield("phone", false)}
	out, errs := validateIntake(fields, IntakeValues{
		Name:  "  Alice Example  ",
		Email: "alice@Example.com",
		Phone: "(555) 123-4567",
	})
	if errs != nil {
		t.Fatalf("expected valid, got %+v", errs)
	}
	if out.Name != "Alice Example" || out.Email != "alice@example.com" || out.Phone != "+15551234567" {
		t.Fatalf("unexpected normalization: %+v", out)
	}
}

func TestValidateIntakeAllErrorsInOnePass(t *testing.T) {
	fields := []IntakeField{ifield("name", true)}
	_, errs := validateIntake(fields, IntakeValues{Email: "not-an-email", Phone: "nope"})
	if len(errs) != 3 {
		t.Fatalf("missing name + bad email + bad phone should be 3 errors, got %+v", errs)
	}
}

func TestValidateIntakeVolunteeredFieldsKept(t *testing.T) {
	// Only email requested, but the visitor volunteered a phone — keep it.
	fields := []IntakeField{ifield("email", true)}
	out, errs := validateIntake(fields, IntakeValues{Email: "a@b.co", Phone: "+15551234567"})
	if errs != nil {
		t.Fatalf("expected valid, got %+v", errs)
	}
	if out.Phone != "+15551234567" {
		t.Fatalf("volunteered phone dropped: %+v", out)
	}
}

func TestIdentityIntakeParseRequestShorthandAndUnknown(t *testing.T) {
	req, err := IdentityIntakeKind{}.ParseRequest(map[string]any{
		"fields": []any{"email", map[string]any{"key": "name", "required": false}},
		"reason": "to send you the quote",
	})
	if err != nil {
		t.Fatalf("parse: %v", err)
	}
	var spec identityIntakeSpec
	if err := json.Unmarshal(req.Spec, &spec); err != nil {
		t.Fatalf("spec unmarshal: %v", err)
	}
	// Shorthand string "email" ⇒ required; explicit object keeps required:false.
	if len(spec.Fields) != 2 || !spec.Fields[0].Required || spec.Fields[1].Required {
		t.Fatalf("unexpected fields: %+v", spec.Fields)
	}
	if _, err := (IdentityIntakeKind{}).ParseRequest(map[string]any{"fields": []any{"ssn"}, "reason": "x"}); err == nil {
		t.Fatal("unknown field key should error")
	}
}

// TestIdentityIntakeFixturesValidate cross-checks the Go kind against the shared conformance
// fixtures the Rust reference validates against: identity_intake_values against
// identity_intake_spec must validate to the identity_intake_payload's normalized values.
func TestIdentityIntakeFixturesValidate(t *testing.T) {
	fixtures := loadFixtures(t)
	spec := fixtureInstance(t, fixtures, "identity_intake_spec")
	values := fixtureInstance(t, fixtures, "identity_intake_values")
	payload := fixtureInstance(t, fixtures, "identity_intake_payload")

	specJSON, err := json.Marshal(spec)
	if err != nil {
		t.Fatalf("marshal spec: %v", err)
	}
	canonical, errs := IdentityIntakeKind{}.Validate(specJSON, values)
	if errs != nil {
		t.Fatalf("shared identity_intake_values should validate against the spec, got %v", errs)
	}
	if !jsonEqual(canonical, payload["values"]) {
		got, _ := json.Marshal(canonical)
		want, _ := json.Marshal(payload["values"])
		t.Fatalf("canonical values != fixture payload values\n got: %s\nwant: %s", got, want)
	}
	if payload["status"] != "submitted" {
		t.Fatalf("fixture payload status = %v, want submitted", payload["status"])
	}
}
